src/libostree/ostree-sign-x509.c

```diff
@@ -0,0 +1,651 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *  - Denis Pynkin (d4s) <denis.pynkin@collabora.com>
+ */
+
+#include "config.h"
+
+#include "ostree-sign-x509.h"
+#include "otcore.h"
+#include <libglnx.h>
+#include <ot-checksum-utils.h>
+#include <string.h>
+
+#undef G_LOG_DOMAIN
+#define G_LOG_DOMAIN "OSTreeSign"
+
+#define OSTREE_SIGN_X509_NAME "x509"
+
+typedef enum
+{
+  X509_OK,
+  X509_NOT_SUPPORTED,
+  X509_FAILED_INITIALIZATION
+} x509_state;
+
+struct _OstreeSignX509
+{
+  GObject parent;
+  x509_state state;
+  GBytes *secret_key;
+  GList *public_keys;  /* GBytes */
+  GList *revoked_keys; /* GBytes */
+};
+
+static void ostree_sign_x509_iface_init (OstreeSignInterface *self);
+
+G_DEFINE_TYPE_WITH_CODE (OstreeSignX509, _ostree_sign_x509, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (OSTREE_TYPE_SIGN, ostree_sign_x509_iface_init));
+
+static void
+ostree_sign_x509_iface_init (OstreeSignInterface *self)
+{
+
+  self->data = ostree_sign_x509_data;
+  self->data_verify = ostree_sign_x509_data_verify;
+  self->get_name = ostree_sign_x509_get_name;
+  self->metadata_key = ostree_sign_x509_metadata_key;
+  self->metadata_format = ostree_sign_x509_metadata_format;
+  self->clear_keys = ostree_sign_x509_clear_keys;
+  self->set_sk = ostree_sign_x509_set_sk;
+  self->set_pk = ostree_sign_x509_set_pk;
+  self->add_pk = ostree_sign_x509_add_pk;
+  self->load_pk = ostree_sign_x509_load_pk;
+}
+
+static void
+_ostree_sign_x509_class_init (OstreeSignX509Class *self)
+{
+}
+
+static void
+_ostree_sign_x509_init (OstreeSignX509 *self)
+{
+
+  self->state = X509_OK;
+  self->secret_key = NULL;
+  self->public_keys = NULL;
+  self->revoked_keys = NULL;
+
+#if !defined(USE_OPENSSL)
+  self->state = X509_NOT_SUPPORTED;
+#else
+  if (!otcore_x509_init ())
+    self->state = X509_FAILED_INITIALIZATION;
+#endif
+}
+
+static gboolean
+_ostree_sign_x509_is_initialized (OstreeSignX509 *self, GError **error)
+{
+  switch (self->state)
+    {
+    case X509_OK:
+      break;
+    case X509_NOT_SUPPORTED:
+      return glnx_throw (error, "x509: engine is not supported");
+    case X509_FAILED_INITIALIZATION:
+      return glnx_throw (error, "x509: crypto library isn't initialized properly");
+    }
+
+  return TRUE;
+}
+
+gboolean
+ostree_sign_x509_data (OstreeSign *self, GBytes *data, GBytes **signature,
+                       GCancellable *cancellable, GError **error)
+{
+
+  g_assert (OSTREE_IS_SIGN (self));
+  OstreeSignX509 *sign = _ostree_sign_x509_get_instance_private (OSTREE_SIGN_X509 (self));
+
+  if (!_ostree_sign_x509_is_initialized (sign, error))
+    return FALSE;
+
+  if (sign->secret_key == NULL)
+    return glnx_throw (error, "Not able to sign: secret key is not set");
+
+#if defined(USE_OPENSSL)
```

Comment:
I'm ok with this as is, but it would seem to me to be cleaner to basically `#ifdef` the whole function and have the `#else` block there be a "return unsupported" error...I think we have an example of that in `otcore_validate_ed25519_signature`.

Reply:
Done.